LibreHardwareMonitorLib/Hardware/Control.cs

```diff
@@ -90,5 +130,331 @@ public void SetSoftware(float value)
+                if (_mode == ControlMode.SoftwareCurve)
+                    AttachSoftwareCurve(_softwareCurve);
+            }
+        }
+    
+        public void NotifyHardwareRemoved(IHardware hardware)
+        {
+            if (_softwareCurve == null)
+                return;
+
+            Debug.WriteLine("notify hardware removed");
+
+            foreach (ISensor sensor in hardware.Sensors)
+                if (sensor.Identifier.ToString() == _sensorIdentifier)
+                {
+                    NotifyClosing();
+                }
+
+            foreach (IHardware subHardware in hardware.SubHardware)
+                NotifyHardwareRemoved(subHardware);
+        }
+
+        public void NotifyClosing()
+        {
+            if (_softwareCurve == null)
+                return;
+
+            DetachSoftwareCurve();
+            _softwareCurve.Dispose();
+            _softwareCurve = null;
+            ControlModeChanged?.Invoke(this);
+            Debug.WriteLine("closing");
+        }
+    
+        private void AttachSoftwareCurve(SoftwareCurve newCurve)
+        {
+            if (_softwareCurveAttached || _softwareCurve != null) DetachSoftwareCurve();
+
+            _softwareCurve = newCurve;
+            //this.softwareCurve.Sensor.Hardware.SensorRemoved += SensorRemoved;
+            //this.parentSensor.Hardware.SensorRemoved += SensorRemoved;
+            _softwareCurve.SoftwareCurveValueChanged += this.HandleSoftwareCurveValueChange;
+            _softwareCurve.SoftwareCurveAbort += this.HandleSoftwareCurveAbort;
+            _softwareCurve.Start();
+            _softwareCurveAttached = true;
+            Debug.WriteLine("attaching softwarecurve");
+        }
+
+        private void DetachSoftwareCurve()
+        {
+            if (!_softwareCurveAttached || _softwareCurve == null) return;
+
+            _softwareCurve.Stop();
+            //this.softwareCurve.Sensor.Hardware.SensorRemoved -= SensorRemoved;
+            //this.parentSensor.Hardware.SensorRemoved -= SensorRemoved;
+            _softwareCurve.SoftwareCurveValueChanged -= this.HandleSoftwareCurveValueChange;
+            _softwareCurve.SoftwareCurveAbort -= this.HandleSoftwareCurveAbort;
+            _softwareCurveAttached = false;
+            Debug.WriteLine("detaching softwarecurve");
+        }
+
+        private void HandleSoftwareCurveValueChange(SoftwareCurve softwareCurve)
+        {
+            _softwareCurveValue = softwareCurve.Value;
+            Debug.WriteLine("setting value from software curve: " + softwareCurve.Value);
+            SoftwareControlValueChanged?.Invoke(this);
+        }
+
+        private void HandleSoftwareCurveAbort(SoftwareCurve softwareCurve)
+        {
+            DetachSoftwareCurve();
+            ControlModeChanged?.Invoke(this); // until softwarecurve is started again, get value of ControlMode is Default
+            Debug.WriteLine("softwarecurve abort!");
+        }
+    }
+    
+    public class SoftwareCurve
```

Comment:
Move to separate file + formatting needs to be fixed.

Reply:
Yep, i'll pull this and the other out into separate files.